### PR TITLE
Add types for interface and links.

### PR DESCRIPTION
**Pull request by howbazaar**
Based on PR #21, this branch adds interfaces and links. These are needed to be able to add static IP addresses to new interfaces for container networking.

The boot interface and interface set are now also exported from Machine.

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -3,175 +3,191 @@
...
-//
-// For example, a distributed network may have a fabric in London containing
-// VLAN 100, while a separate fabric in San Francisco may contain a VLAN 100,
-// whose attached subnets are completely different and unrelated.
-type Fabric interface {
-	ID() int
-	Name() string
-	ClassType() string
+// Can't use interface as a type, so add an underscore. Yay.
+type interface_ struct {
```
Comment (mjs) [inline interface.go:13]: sad trombone :)

File: interfaces.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+// Controller represents an API connection to a MAAS Controller. Since the API
+// is restful, there is no long held connection to the API server, but instead
+// HTTP calls are made and JSON response structures parsed.
+type Controller interface {
+
+	// Capabilities returns a set of capabilities as defined by the string
+	// constants.
+	Capabilities() set.Strings
+
+	BootResources() ([]BootResource, error)
```
Comment (mjs) [inline interfaces.go:27]: Missing docstring

File: interfaces.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	// Spaces returns the list of Spaces defined in the MAAS controller.
+	Spaces() ([]Space, error)
+
+	// Zones lists all the zones known to the MAAS controller.
+	Zones() ([]Zone, error)
+
+	// Machines returns a list of machines that match the params.
+	Machines(MachinesArgs) ([]Machine, error)
+
+	AllocateMachine(AllocateMachineArgs) (Machine, error)
```
Comment (mjs) [inline interfaces.go:43]: Docstrings

File: interfaces.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	ID() int
+	Name() string
+	Fabric() string
+
+	// VID is the VLAN ID. eth0.10 -> VID = 10.
+	VID() int
+	// MTU (maximum transmission unit) is the largest size packet or frame,
+	// specified in octets (eight-bit bytes), that can be sent.
+	MTU() int
+	DHCP() bool
```
Comment (mjs) [inline interfaces.go:97]: Docstrings here?

File: interfaces.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+// Zone represents a physical zone that a Machine is in. The meaning of a
+// physical zone is up to you: it could identify e.g. a server rack, a network,
+// or a data centre. Users can then allocate nodes from specific physical zones,
+// to suit their redundancy or performance requirements.
+type Zone interface {
+	Name() string
+	Description() string
+}
+
+// BootResource is the bomb... find something to say here.
```
Comment (mjs) [inline interfaces.go:112]: Lots of docstrings missing from here on

File: link.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/juju/version"
+)
+
+type link struct {
+	id     int
+	mode   string
+	subnet *subnet
+}
+
+// NOTE: not using lowercase L as the receiver as it is a horrible idea.
```
Comment (mjs) [inline link.go:18]: FWIW, Will often reuses the type name. For example:

``` go
func (link *link) ID() int {
```

YMMV

**Comment by mjs** [comment]
LGTM. Just missing docstrings.

File: link.go (hunk 6 above)
Comment (howbazaar) [inline link.go:18]: I considered it, but just went with a traditional one character receiver.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**